### ISSUE TITLE
Fixed-size consensus types should have `from_consensus_bytes(bytes: &[u8; N])`

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-bitcoin#1458]

**Comment by PaarthAgarwal** [comment]
Can you provide more context for this issue?

**Comment by apoelstra** [comment]
@PaarthAgarwal almost all types in Bitcoin have a well-defined fixed-size "consensus encoding". The only exceptions are vectors/slices which are variable sized, and varints (which maybe aren't actually a distinct type anymore? I forget).

So for everything that implements `consensus::Decodable`, except for the exceptions, we should have a method like `deserialize_fixed([u8; 4])`. Probably this should be in a new trait, `consensus::DecodableFixed`.

**Comment by Kixunil** [comment]
I think trait is too much. Unless I'm missing some interesting use case. Also trait would look nicer after MSRV bump.

**Comment by apoelstra** [comment]
No, I have no real use case. I'd be happy without the trait. Good observation that our current MSRV would make it kinda ugly anyway.

**Comment by liamaharon** [comment]
Hey @apoelstra, I'm interested to take this one on.

> almost all types in Bitcoin have a well-defined fixed-size "consensus encoding"

Do you mean a standard across implementations? Is there a spec I can refer to?

**Comment by apoelstra** [comment]
All Bitcoin consensus rules are defined with respect to the reference implementation at https://github.com/bitcoin/bitcoin/tree/master/src/

**Comment by Dmenec** [comment]
Hi, I’d like to contribute to this if it’s still needed. Should I go with implementing individual functions on each type or introduce a trait?

**Comment by apoelstra** [comment]
So, for many types we already have (essentially) these methods, with names like `from_consensus_u32`. This is true, AFAIK, for:

* all types which are wrappers around 32-bit uints (if you have bytes you can start by calling `u32::from_le_bytes` which is infallible)
* all hash types (which have `from_byte_array`)

Then for composite types like `OutPoint` which are a 32-byte array next to a u32, we have infallible methods to split up the array so you can still do this infallibly. See e.g. #5287. It feels a bit niche to me to provide a convenience method for this.

I don't think this makes sense as a trait because it would need to be tied to the integer length somehow, and it's not clear to me when it would be useful to be generic over that.

So maybe we can close this issue as "complete"?

**Comment by tcharding** [comment]
> all types which are wrappers around 32-bit uints

Also applies to signed int wrapper types e.g. `block::Version`.

> So maybe we can close this issue as "complete"?

I looked around for five minutes or so and I believe we can.
